src/hotspot/share/opto/reachability.cpp

```diff
@@ -0,0 +1,555 @@
+ *       safepoint nodes (appended as edges after debug info);
+ *   (3) during final graph reshaping, referent edges are removed from safepoints and materialized as RF nodes
+ *       attached to their safepoint node (closely following it in CFG graph).
+ *
+ * Some implementation considerations.
+ *
+ * (a) It looks attractive to get rid of RF nodes early and transfer to safepoint-attached representation,
+ * but it is not correct until loop opts are done.
+ *
+ * Live ranges of values are routinely extended during loop opts. And it can break the invariant that
+ * all interfering safepoints contain the referent in their oop map. (If an interfering safepoint doesn't
+ * keep the referent alive, then it becomes possible for the referent to be prematurely GCed.)
+ *
+ * After loop opts are over, it becomes possible to reliably enumerate all interfering safe points and
+ * to ensure that the referent is present in their oop maps.
+ *
+ * (b) RF nodes may interfere with Register Allocator (RA). If a safepoint is pruned during macro expansion,
+ * it can make some RF nodes redundant, but we don't have information about their relations anymore to detect that.
+ * Redundant RF node unnecessarily extends referent's live range and increases register pressure.
+ *
+ * Hence, we eliminate RF nodes and transfer their referents to corresponding safepoints (phase #2).
+ * When safepoints are pruned, corresponding reachability edges also go away.
+ *
+ * (c) Unfortunately, it's not straightforward to stay with safepoint-attached representation till the very end,
+ * because information about derived oops is attached to safepoints in a similar way. So, for now RFs are
+ * rematerialized at safepoints before RA (phase #3).
+ */
+
+// RF is redundant for some referent oop when the referent has another user which keeps it alive across the RF.
+// In terms of dominance relation it can be formulated as "a referent has a user which is dominated by the redundant RF".
+// Until loop opts are over, only RF nodes are considered as usages (controlled by rf_only flag).
+static bool is_redundant_rf_helper(ReachabilityFenceNode* rf, PhaseIdealLoop* phase, PhaseGVN& gvn, bool rf_only) {
+  assert(phase != nullptr || rf_only, "only RFs during GVN");
+
+  Node* referent = rf->referent();
+  const Type* t = gvn.type(referent);
+  if (t == TypePtr::NULL_PTR) {
+    return true; // no-op fence: null referent
+  }
+  if (!OptimizeReachabilityFences) {
+    return false; // keep reachability fence nodes intact
+  }
+  if (!PreserveReachabilityFencesOnConstants && t->singleton()) {
+    return true; // no-op fence: constants are strongly reachable
+  }
+  if (referent->is_Proj() && referent->in(0)->is_CallJava()) {
+    ciMethod* m = referent->in(0)->as_CallJava()->method();
+    if (m != nullptr && m->is_boxing_method()) {
+      return true; // ignore fences on boxed primitives produced by valueOf methods
+    }
+  }
```

Comment:
Nice, thanks for adding the comment. I'm trying to understand the reason why that is ok.

So someone would have set a RF for a boxed primitive. But we don't expect anything to ever be attached to a boxed primitive, and so we can just ignore these RF? Is that the reason? Might be worth writing it in a code comment explicitly.

Reply:
I stumbled upon it while inspecting IR with `-XX:+StressReachabiltiyFences` when RF is added for auto-boxed argument.  (`is_boxing_method()` matches `valueOf` factory, not an explicitly allocated box).

Thinking more about it, there's a way to observe the absence of RF on a auto-boxed instance when (1) value is out of range for internal caches; and (2) there's java.lang.ref.Reference instance registered for it.

So, I moved the logic under `StressReachabilityFences` instead.